ansible/roles/dev/tasks/main.yml

```diff
@@ -70,40 +56,138 @@
   user:
-      name: vagrant
+      name: "{{ ansible_user }}"
       state: present
       groups: systemd-journal
       append: true
 
-- name: allow vagrant user to use mock
+- name: Ensure unprivileged user can use mock
   user:
-      name: vagrant
+      name: "{{ ansible_user }}"
       state: present
       groups: mock
       append: true
 
 - name: Install Pulp RPM dependencies
-  dnf: name={{ item }} state=present
+  package: name={{ item }} state=present
   with_items: "{{ pulp_rpm_dependencies }}"
 
-- name: Ensure homedir mode suitable for ssh keys
-  file: dest=/home/{{ ansible_env.SUDO_USER }} mode=0750
+- block:
+  # block acts as unprivileged user to set up virtualenvs and install packages
+  - name: Ensure virtualenvs dir exists
+    file:
+        path: "{{ pulp_venv_dir }}"
+        state: directory
+
+  - name: Create virtualenvs
+    # the python3 venv module does not install pip if pip3 is installed in the system and
+    # you specify --system-site-packages, so we have to use virtualenv
+    command: "virtualenv --python /usr/bin/python3 --system-site-packages {{ pulp_venv_dir }}/{{ item }}"
+    args:
+        creates: "{{ pulp_venv_dir }}/{{ item }}"
+    with_items: "{{ pulp_available_repositories | difference(pulp_created_virtualenvs) }}"
+
+  - name: Link virtualenv project dirs
+    lineinfile:
+        dest: "{{ pulp_venv_dir }}/{{ item }}/.project"
+        line: "{{ pulp_devel_dir }}/{{ item }}"
+        create: true
+        regexp: ".*"
+    with_items: "{{ pulp_available_repositories }}"
+
+  - name: Install project requirements
+    pip:
+    args:
+        virtualenv: "{{ pulp_venv_dir }}/{{ item.key }}"
+        state: latest
+        chdir: "{{ pulp_devel_dir }}/{{ item.key }}"
+        requirements: "{{ item.value | join(' -r ') }}"
+    vars:
+      # mapping of repository: ['list', 'of', 'requirements'] to install into the dev environment
+      # in addition to the "actual" python requirements defined in the various setup.py files.
+      requirements_files:
+          pulp:
+              - test_requirements.txt
+          crane:
+              - test-requirements.txt
+          pulp-smash:
+              - requirements.txt
+              - requirements-dev.txt
+    with_dict: "{{ requirements_files }}"
+    # only install requirements for projects that are available :)
+    when: "{{ [item.key] | issubset(pulp_available_repositories) }}"
+
+  - name: Install Pulp Platform packages
+    command: "{{ pulp_venv_dir }}/pulp/bin/python setup.py develop"
+    args:
+      chdir: "{{ pulp_devel_dir }}/pulp/{{ item }}"
+    # Pulp 2 setup.py files still exist on 3.0-dev in platform, so we can't just find
+    # every setup.py and run develop on it with the virtualenv interpreter. This is also true
+    # for plugins, in which we'll likely want to install all plugin setup.py as well as
+    # all platform setup.pys for each plugin virtualenv.
+    # TODO: Make a redmine task to track this
```

Comment:
I think https://pulp.plan.io/issues/278 covers this nicely, so I can just delete the TODO and add a reference to that issue